### PR TITLE
Add more type hints

### PR DESCRIPTION
**Pull request by core23**
## Subject

This is a follow up of #6351 to add more type hints.

I am targeting this branch, because this is BC.

## Changelog

```markdown
### Added
- Added generic type hints to (nearly) all admin interfaces

### Fix
-  Fix calling undefined `AdminInterface::isCurrentChild` 
```

### CODE REVIEW

File: src/Admin/AdminInterface.php
Comment (VincentLanglet) [inline src/Admin/AdminInterface.php]: @phpstan-return ?

File: src/Admin/AdminInterface.php
Comment (VincentLanglet) [inline src/Admin/AdminInterface.php:644]: You can add `phpstan-param array<class-string>`.

File: src/Admin/AdminExtensionInterface.php
Comment (VincentLanglet) [inline src/Admin/AdminExtensionInterface.php:129]: I saw you removed lot of `@return void`.

For phpstan `@return mixed` is the default behavior. So `@return void` is needed to have a better analysis.

**Review by VincentLanglet — COMMENTED** [review]
I have trouble calling this as BC.

For instance, someone with a phpstan level > 6 will get an error if he use `AdminInterface` instead of `AdminInterface<object>` as a param/return type.

It should be at least minor.

File: src/Admin/AbstractAdmin.php
Comment (VincentLanglet) [inline src/Admin/AbstractAdmin.php:2997]: T|null

File: src/Admin/AbstractAdmin.php
Comment (VincentLanglet) [inline src/Admin/AbstractAdmin.php:3109]: T|null

File: src/Admin/AdminInterface.php
Comment (VincentLanglet) [inline src/Admin/AdminInterface.php:115]: phpstan-return ?

File: src/Admin/AdminExtensionInterface.php
Comment (core23) [reply src/Admin/AdminExtensionInterface.php:129]: Are you sure? Do you have a link where this is explained?

Comment (VincentLanglet) [reply src/Admin/AdminExtensionInterface.php:129]: https://github.com/phpstan/phpstan/issues/2842#issuecomment-573139424

**Review by VincentLanglet — APPROVED** [review]
Maybe @greg0ire or @vladyslavstartsev can help for the psalm issues.

**Comment by vladyslavstartsev** [comment]
By the way. This can be one of multiple PRs to get PHPStan level6. You can temporarily enable it (without committing).
There is 956 errors on level 6 currently.
If you do level 6 + `checkMissingIterableValueType: false` + `checkGenericClassInNonGenericObjectType: false` there is 495 errors.

**Comment by VincentLanglet** [comment]
I don't know where come from the error with the toString part.
There is no issue on psalm sandbox https://psalm.dev/r/d4d8ee7902

**Comment by core23** [comment]
Adding type hints to `AdminExtensionInterface` looks complicated, so I removed them for now

**Comment by VincentLanglet** [comment]
If you rebase 3.x i think you can do the same for getNormalizeIdentifier et getUrlSafeIdentifier

**Comment by core23** [comment]
> getUrlSafeIdentifier

I don't understand. This PR is already rebased.

Did you mean `ModelManagerInterface::getNormalizedIdentifier` and `ModelManagerInterface::getUrlSafeIdentifier`? This interface can not be used with general type hints, because there is just an instance for all admins.

**Comment by VincentLanglet** [comment]
> Did you mean `ModelManagerInterface::getNormalizedIdentifier` and `ModelManagerInterface::getUrlSafeIdentifier`? This interface can not be used with general type hints, because there is just an instance for all admins.

You can use 
```
@phpstan-param T $model
```
for AdminInterface::getNormalizedIdentifier.

Same for UrlGeneratorInterface::getUrlSafeIdentifier.

File: src/Admin/AbstractAdmin.php
Comment (core23) [inline src/Admin/AbstractAdmin.php:2427]: The old method does not exist on the interface

**Comment by core23** [comment]
I found a bug when adding these type hints, because we are calling a method that must not exist.

The most common methods are now type hinted. The global type coverage is around ~85%.

Please review again @VincentLanglet

**Comment by VincentLanglet** [comment]
> The global type coverage is around ~85%.

How do you know this ?

**Review by VincentLanglet — DISMISSED** [review]

**Comment by core23** [comment]
> How do you know this ?

```bash
psalm --stats
```

Refs: https://github.com/sonata-project/dev-kit/issues/867

File: src/Admin/BreadcrumbsBuilderInterface.php
Comment (greg0ire) [inline src/Admin/BreadcrumbsBuilderInterface.php:33]: We don't actually require implementations to return an array, do we?

```suggestion
     * @return iterable<ItemInterface> the breadcrumbs
```

File: src/Filter/Persister/FilterPersisterInterface.php
Comment (greg0ire) [inline src/Filter/Persister/FilterPersisterInterface.php:46]: Could be great to have an array shape here, have you considered it?

File: src/Admin/BreadcrumbsBuilderInterface.php
Comment (core23) [reply src/Admin/BreadcrumbsBuilderInterface.php:33]: `BreadcrumbsBuilder::getBreadcrumbs` returns an array. There is no usage of `yields` in our code, so we should use simple arrays.

File: src/Filter/Persister/FilterPersisterInterface.php
Comment (core23) [reply src/Filter/Persister/FilterPersisterInterface.php:46]: You can add custom array keys here, so we can't enforce them

**Review by VincentLanglet — DISMISSED** [review]

**Review by VincentLanglet — APPROVED** [review]

**Comment by core23** [comment]
I backported your type hints from https://github.com/sonata-project/SonataAdminBundle/pull/6267. 

Please review @jordisala1991

**Review by VincentLanglet — APPROVED** [review]

File: src/Datagrid/Pager.php
Comment (jordisala1991) [inline src/Datagrid/Pager.php:574]: no need to use phpstan-return here (all this file)

File: src/Admin/Pool.php
Comment (jordisala1991) [inline src/Admin/Pool.php:201]:
```
@phpstan-template T
@phpstan-param class-string<T> $class
@phpstan-return AdminInterface<T>|null
```

File: src/Form/DataTransformer/ModelToIdPropertyTransformer.php
Comment (jordisala1991) [inline src/Form/DataTransformer/ModelToIdPropertyTransformer.php:67]: or null

File: src/Form/DataTransformer/ModelToIdPropertyTransformer.php
Comment (jordisala1991) [inline src/Form/DataTransformer/ModelToIdPropertyTransformer.php:123]: This param should be T[]|T|null?

Comment (jordisala1991) [reply src/Form/DataTransformer/ModelToIdPropertyTransformer.php:123]: Not sure if it has to be a collection, the code seems to say it could be only an array

File: src/Security/Handler/AclSecurityHandler.php
Comment (jordisala1991) [inline src/Security/Handler/AclSecurityHandler.php:191]: Why you removed all the information?

File: src/Templating/TemplateRegistry.php
Comment (jordisala1991) [inline src/Templating/TemplateRegistry.php:89]: why you remove this one?

File: src/Security/Handler/AclSecurityHandler.php
Comment (core23) [reply src/Security/Handler/AclSecurityHandler.php:191]: The same comment exists on the interface

File: src/Templating/TemplateRegistry.php
Comment (core23) [reply src/Templating/TemplateRegistry.php:89]: The same comment exists on the interface

**Review by VincentLanglet — APPROVED** [review]

File: src/Admin/Pool.php
Comment (core23) [reply src/Admin/Pool.php]: phpstan got some troubles to resolve this correctly. So I added this to the baseline

File: src/Admin/Pool.php
Comment (jordisala1991) [inline src/Admin/Pool.php:203]: Maybe using template T of object makes the psalm error go away?

File: src/Form/DataTransformer/ModelsToArrayTransformer.php
Comment (jordisala1991) [inline src/Form/DataTransformer/ModelsToArrayTransformer.php:136]: The collection is from T elements? Maybe you can add that to the phpdoc

**Comment by jordisala1991** [comment]
This one is almost done, but for next PRs doing a backport, it would be nice to do them after the main PR is merged, otherwise I have to review again my work, and when merging back to master, you will have to review again how is that applied to my PR

**Comment by core23** [comment]
Green again :)

There may be some more type hints that could be added, but this can be done in a separate PR.

File: src/Admin/BreadcrumbsBuilderInterface.php
Comment (greg0ire) [reply src/Admin/BreadcrumbsBuilderInterface.php:33]: That should not be dictated by our implementations, but by our usage in the client code, which doesn't require the result to be an array AFAIK

File: .phpstan/phpstan-baseline.neon
Comment (VincentLanglet) [inline .phpstan/phpstan-baseline.neon:49]: Just saw you added this.

Shouldn't we this instead ?
- Either the phpdoc is wrong
- Either the phpdoc is right and the check is useless. Then we should decide if we remove this in 3.x or in the next major (and so, a NEXT_MAJOR should be added).

Comment (core23) [reply .phpstan/phpstan-baseline.neon:49]: I won't trust the PHPDoc in this case, because a transformer is mostly used in a *magic* form context which does not use a strict typing system.

The code verifies, if you really pass the correct type.

Comment (VincentLanglet) [reply .phpstan/phpstan-baseline.neon:49]: I don't understand. If you don't trust the phpdoc, then change it.

The Phpdoc from Symfony is saying 
```
    /**
     * This method must be able to deal with empty values. Usually this will
     * be NULL, but depending on your implementation other empty values are
     * possible as well (such as empty strings). The reasoning behind this is
     * that data transformers must be chainable. If the transform() method
     * of the first data transformer outputs NULL, the second must be able to
     * process that value.
     *
     * @param mixed $value The value in the original representation
     *
     * @return mixed The value in the transformed representation
     *
     * @throws TransformationFailedException when the transformation fails
     */
    public function transform($value);
```

So handling `null` value seems the right things to do. You should write
```
    /**
     * @param obejct[]|null $collection
     *
     * @return string[]
     *
     * @phpstan-param T[]|null $collection
     */
    public function transform($collection)
```
And you won't get the error.

Comment (core23) [reply .phpstan/phpstan-baseline.neon]: I reverted these things and will create a new PR for the mapper components. This might result in a longer discussion ;)

**Review by VincentLanglet — APPROVED** [review]

**Review by jordisala1991 — APPROVED** [review]

**Pull request merged**
